File: assets/codebase/menu/MenuDefinition.cs
namespace city.menu {
    /// <summary>
    /// Captures the complete city menu host configuration including panels, theme colors, and fonts.
    /// </summary>
    public sealed class MenuDefinition {
        /// <summary>
        /// Initializes one city menu definition.
        /// </summary>
        /// <param name="title">Optional large display title rendered near the top of the menu.</param>
        /// <param name="subtitle">Optional supporting subtitle rendered under the title.</param>
        /// <param name="initialPanelId">Panel id that should become active when the menu host starts.</param>
        /// <param name="titleFontPath">Project-relative path to the title font asset.</param>
        /// <param name="bodyFontPath">Project-relative path to the body font asset.</param>
        /// <param name="backgroundColor">Primary background color used by the menu canvas.</param>
        /// <param name="surfaceColor">Surface color used by panel chrome.</param>
        /// <param name="surfaceBorderColor">Border color used by panel chrome.</param>
        /// <param name="accentColor">Primary accent color used by focused and hovered elements.</param>
        /// <param name="accentSecondaryColor">Secondary accent color used by decorative accents.</param>
        /// <param name="textColor">Primary text color used for headings and labels.</param>
        /// <param name="mutedTextColor">Secondary text color used for subtitles and descriptions.</param>
        /// <param name="panels">Panels available to the menu host.</param>
        /// <param name="overlayImage">Optional decorative overlay image baked into the menu canvas.</param>
        /// <param name="platformInfoOverlay">Optional top-right platform information overlay baked into the menu canvas.</param>
        public MenuDefinition(
            string title,
            string subtitle,
            string initialPanelId,
            string titleFontPath,
            string bodyFontPath,
            byte4 backgroundColor,
            byte4 surfaceColor,
            byte4 surfaceBorderColor,
            byte4 accentColor,
            byte4 accentSecondaryColor,
            byte4 textColor,
            byte4 mutedTextColor,
            MenuPanelDefinition[] panels,
            MenuOverlayImageDefinition overlayImage = null,
            MenuPlatformInfoDefinition platformInfoOverlay = null) {
            if (title == null) {
                throw new ArgumentNullException(nameof(title));
            }
            if (subtitle == null) {
                throw new ArgumentNullException(nameof(subtitle));
            }
            if (string.IsNullOrWhiteSpace(initialPanelId)) {
                throw new ArgumentException("Initial panel id must be provided.", nameof(initialPanelId));
            }
            if (string.IsNullOrWhiteSpace(titleFontPath)) {
                throw new ArgumentException("Title font path must be provided.", nameof(titleFontPath));
            }
            if (string.IsNullOrWhiteSpace(bodyFontPath)) {
                throw new ArgumentException("Body font path must be provided.", nameof(bodyFontPath));
            }
            if (panels == null) {
                throw new ArgumentNullException(nameof(panels));
            }
            if (panels.Length == 0) {
                throw new InvalidOperationException("Menu definitions must contain at least one panel.");
            }

            Title = title;
            Subtitle = subtitle;
            InitialPanelId = initialPanelId;
            TitleFontPath = titleFontPath;
            BodyFontPath = bodyFontPath;
            BackgroundColor = backgroundColor;
            SurfaceColor = surfaceColor;
            SurfaceBorderColor = surfaceBorderColor;
            AccentColor = accentColor;
            AccentSecondaryColor = accentSecondaryColor;
            TextColor = textColor;
            MutedTextColor = mutedTextColor;
            Panels = panels;
            OverlayImage = overlayImage;
            PlatformInfoOverlay = platformInfoOverlay;
        }

        /// <summary>
        /// Gets the optional large display title rendered near the top of the menu.
        /// </summary>
        public string Title { get; }

        /// <summary>
        /// Gets the optional supporting subtitle rendered under the title.
        /// </summary>
        public string Subtitle { get; }

        /// <summary>
        /// Gets the panel id activated when the menu host starts.
        /// </summary>
        public string InitialPanelId { get; }

        /// <summary>
        /// Gets the project-relative path to the title font asset.
        /// </summary>
        public string TitleFontPath { get; }

        /// <summary>
        /// Gets the project-relative path to the body font asset.
        /// </summary>
        public string BodyFontPath { get; }

        /// <summary>
        /// Gets the primary background color used by the menu canvas.
        /// </summary>
        public byte4 BackgroundColor { get; }

        /// <summary>
        /// Gets the surface color used by panel chrome.
        /// </summary>
        public byte4 SurfaceColor { get; }

        /// <summary>
        /// Gets the border color used by panel chrome.
        /// </summary>
        public byte4 SurfaceBorderColor { get; }

        /// <summary>
        /// Gets the primary accent color used by focused and hovered elements.
        /// </summary>
        public byte4 AccentColor { get; }

        /// <summary>
        /// Gets the secondary accent color used by decorative accents.
        /// </summary>
        public byte4 AccentSecondaryColor { get; }

        /// <summary>
        /// Gets the primary text color used for headings and labels.
        /// </summary>
        public byte4 TextColor { get; }

        /// <summary>
        /// Gets the secondary text color used for supporting copy.
        /// </summary>
        public byte4 MutedTextColor { get; }

        /// <summary>
        /// Gets the panels available to the menu host.
        /// </summary>
        public MenuPanelDefinition[] Panels { get; }

        /// <summary>
        /// Gets the optional decorative overlay image baked into the menu canvas.
        /// </summary>
        public MenuOverlayImageDefinition OverlayImage { get; }

        /// <summary>
        /// Gets the optional top-right platform information overlay baked into the menu canvas.
        /// </summary>
        public MenuPlatformInfoDefinition PlatformInfoOverlay { get; }
    }
}

File: windows-build-20260602-directional-shadow-plaza-direct/cooked/code/gameplay/MenuDefinition.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "MenuDefinition.hpp"
#include "runtime/native_string.hpp"
#include "MenuOverlayImageDefinition.hpp"
#include "runtime/array.hpp"
#include "MenuPanelDefinition.hpp"
#include "MenuPlatformInfoDefinition.hpp"
#include "runtime/native_exceptions.hpp"
#include "MenuDefinition.hpp"
#include "runtime/array.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_string.hpp"

byte4 MenuDefinition::get_AccentColor()
{
return this->AccentColor;
}

byte4 MenuDefinition::get_AccentSecondaryColor()
{
return this->AccentSecondaryColor;
}

byte4 MenuDefinition::get_BackgroundColor()
{
return this->BackgroundColor;
}

const std::string& MenuDefinition::get_BodyFontPath()
{
return this->BodyFontPath;
}

const std::string& MenuDefinition::get_InitialPanelId()
{
return this->InitialPanelId;
}

byte4 MenuDefinition::get_MutedTextColor()
{
return this->MutedTextColor;
}

::MenuOverlayImageDefinition* MenuDefinition::get_OverlayImage()
{
return this->OverlayImage;
}

Array<::MenuPanelDefinition*>* MenuDefinition::get_Panels()
{
return this->Panels;
}

::MenuPlatformInfoDefinition* MenuDefinition::get_PlatformInfoOverlay()
{
return this->PlatformInfoOverlay;
}

const std::string& MenuDefinition::get_Subtitle()
{
return this->Subtitle;
}

byte4 MenuDefinition::get_SurfaceBorderColor()
{
return this->SurfaceBorderColor;
}

byte4 MenuDefinition::get_SurfaceColor()
{
return this->SurfaceColor;
}

byte4 MenuDefinition::get_TextColor()
{
return this->TextColor;
}

const std::string& MenuDefinition::get_Title()
{
return this->Title;
}

const std::string& MenuDefinition::get_TitleFontPath()
{
return this->TitleFontPath;
}

MenuDefinition::MenuDefinition(std::string title, std::string subtitle, std::string initialPanelId, std::string titleFontPath, std::string bodyFontPath, byte4 backgroundColor, byte4 surfaceColor, byte4 surfaceBorderColor, byte4 accentColor, byte4 accentSecondaryColor, byte4 textColor, byte4 mutedTextColor, Array<::MenuPanelDefinition*>* panels, ::MenuOverlayImageDefinition* overlayImage, ::MenuPlatformInfoDefinition* platformInfoOverlay) : AccentColor(), AccentSecondaryColor(), BackgroundColor(), BodyFontPath(), InitialPanelId(), MutedTextColor(), OverlayImage(), Panels(), PlatformInfoOverlay(), Subtitle(), SurfaceBorderColor(), SurfaceColor(), TextColor(), Title(), TitleFontPath()
{
    if (String::IsNullOrEmpty(title))
    {
throw new ArgumentNullException("title");
    }
    if (String::IsNullOrEmpty(subtitle))
    {
throw new ArgumentNullException("subtitle");
    }
    if (String::IsNullOrWhiteSpace(initialPanelId))
    {
throw ([&]() {
auto __ctor_arg_00000030 = "Initial panel id must be provided.";
auto __ctor_arg_00000031 = "initialPanelId";
return new ArgumentException(__ctor_arg_00000030, __ctor_arg_00000031);
})();
    }
    if (String::IsNullOrWhiteSpace(titleFontPath))
    {
throw ([&]() {
auto __ctor_arg_00000032 = "Title font path must be provided.";
auto __ctor_arg_00000033 = "titleFontPath";
return new ArgumentException(__ctor_arg_00000032, __ctor_arg_00000033);
})();
    }
    if (String::IsNullOrWhiteSpace(bodyFontPath))
    {
throw ([&]() {
auto __ctor_arg_00000034 = "Body font path must be provided.";
auto __ctor_arg_00000035 = "bodyFontPath";
return new ArgumentException(__ctor_arg_00000034, __ctor_arg_00000035);
})();
    }
    if (panels == nullptr)
    {
throw new ArgumentNullException("panels");
    }
    if (panels->get_Length() == 0)
    {
throw new InvalidOperationException("Menu definitions must contain at least one panel.");
    }
this->Title = title;
this->Subtitle = subtitle;
this->InitialPanelId = initialPanelId;
this->TitleFontPath = titleFontPath;
this->BodyFontPath = bodyFontPath;
this->BackgroundColor = backgroundColor;
this->SurfaceColor = surfaceColor;
this->SurfaceBorderColor = surfaceBorderColor;
this->AccentColor = accentColor;
this->AccentSecondaryColor = accentSecondaryColor;
this->TextColor = textColor;
this->MutedTextColor = mutedTextColor;
this->Panels = panels;
this->OverlayImage = overlayImage;
this->PlatformInfoOverlay = platformInfoOverlay;
}

